rpc-broker/src/policy.c

```diff
@@ -0,0 +1,258 @@
+*/
+
+#include "rpc-broker.h"
+
+
+static int create_rule(struct rule *current, char *rule)
+{
+    char *token;
+    const char *delimiter;
+
+    if (!rule || rule[0] == '\0')
+        return -1;
+
+    current->rule_string = strdup(rule);
+
+    delimiter = " ";
+    token = strtok(rule, delimiter);
+
+    if (!token)
+        return -1;
+
+    current->policy = token[0] == 'a';
+    token = strtok(NULL, delimiter);
+
+    while (token) {
+
+        char *field = strtok(NULL, delimiter);
+
+        if (!field && token[0] != 's') {
+            return -1;
+        } else if (!strcmp("destination", token)) {
+            current->destination = strdup(field);
+        } else if (!strcmp("dom-type", token)) {
+            current->domtype = strdup(field);
+        } else if (!strcmp("interface", token)) {
+            current->interface = strdup(field);
+        } else if (!strcmp("if-boolean", token)) {
+            current->if_bool = strdup(field);
+            token = strtok(NULL, delimiter);
+            current->if_bool_flag = strcmp("true", token) == 0 ? 1 : 0;
+        } else if (!strcmp("stubdom", token)) {
+            current->stubdom = 1;
+        } else if (!strcmp("path", token)) {
+            current->path = strdup(field);
+        } else if (!strcmp("member", token)) {
+            current->member = strdup(field);
+        } else if (!strcmp("out-any", token)) {
+            current->out = 1;
+        } else {
+            DBUS_BROKER_WARNING("Unrecognized Rule-Token: %s", token);
+            free_rule(*current);
+            current = NULL;
+            return -1;
+        }
+
+        token = strtok(NULL, delimiter);
+    }
+
+    return 0;
+}
+
+static inline void get_rules(DBusConnection *conn, struct domain_policy *dom)
+{
+    int rule_idx;
+    char *rulestring;
+    char *arg;
+    struct rule *policy_rule;
+
+    for (rule_idx=0; rule_idx < MAX_RULES; rule_idx++) {
+        DBUS_REQ_ARG(arg, "/vm/%s/rpc-firewall-rules/%d",
+                     dom->uuid, rule_idx);
+
+        rulestring = db_query(conn, arg);
+        free(arg);
+
+        if (!rulestring)
+            break;
+
+        policy_rule = &(dom->rules[rule_idx]);
+
+        if (create_rule(policy_rule, rulestring) < 0)
+            free_rule(*policy_rule);
+        else
+            dom->count++;
+
+        free(rulestring);
+    }
+
+}
+
+static void build_etc_policy(struct etc_policy *etc, const char *rule_filepath)
+{
+    int rule_idx;
+    FILE *policy_fh;
+    size_t rbytes, line_length;
+    char *line;
+    char current_rule[RULE_MAX_LENGTH] = { 0 };
+    struct rule *current;
+
+    rule_idx = 0;
+    etc->count = 0;
+
+    policy_fh = fopen(rule_filepath, "r");
+    if (!policy_fh) {
+        DBUS_BROKER_WARNING("/etc policy stat of file <%s> failed %s",
+                             rule_filepath, strerror(errno));
+        return;
+    }
+
+    line = NULL;
+
+    while ((getline(&line, &rbytes, policy_fh) > 0) && rule_idx < MAX_RULES) {
```

Comment:
So if you've reached MAX_RULES you read a line for nothing. Maybe swap the tests.